Make the phrase "local directory" (and other drive type descriptions) translatable

Wengier opened the pull request:
It was reported in #1922 by @Burrito78 that phrases like "local directory" were not translatable. This PR intends to solve this by making all drive type descriptions (including "CDRom", "fatDrive", "isoDrive", and "Internal Virtual Drive") translatable.

Burrito78 commented:
Hi, thanks @Wengier for taking this on!

I tried this build https://github.com/dosbox-staging/dosbox-staging/actions/runs/3103355417 but couldn't find "local directory" in de.lng. Am i doing something wrong?

Wengier commented:
@Burrito78 I am apparently not a German speaker, so I could only added/updated English language messages. Go ahead to add the translated messages to the German language file too.

Burrito78 commented:
Yes, i've seen it now in en.lng that you added the following strings:

![Bildschirmfoto 2022-09-22 um 10 36 48](https://user-images.githubusercontent.com/61995128/191699799-4fb756f8-664f-4c02-8d78-7702f564e927.png)

kcgen reviewed (changes requested):
The drive types should be defined in an enum, and each drive info object populated with the relevant enum, and then those enums can be compared, as opposed to furthering the previous approach of using string comparisons (and now more of the same, but also using strings from the messaging system).

Burrito78 commented:
OK, it works. Thanks for implementing this!

Burrito78 commented:
@Wengier Why are these phrases so strange? 

Why is it CDrom and not CD-ROM?
fatDrive and not FAT-Drive?
isoDrive and not ISO-Drive?
"Internal Virtual Drive" insteant of "internal virtual drive"

I'm asking because i need to translate these and need to understand how they are utilized.

Wengier commented:
@Burrito78 The phrases are a 1:1 replacement of the original ones as used by DOSBox Staging (or SVN). But given that they can now be changed, I fully agree that better (default) phrases can be used when desired.

Burrito78 commented:
@Wengier Understood and agreed. Can you change them in the english language file?

Wengier commented:
@kcgen I have changed the drive types to enum as suggested.

@Burrito78 I have updated the English language file, which hopefully reads better:

``:MOUNT_TYPE_LOCAL_DIRECTORY``
``local directory``
``.``
``:MOUNT_TYPE_CDROM``
``CD-ROM drive``
``.``
``:MOUNT_TYPE_FAT``
``FAT drive``
``.``
``:MOUNT_TYPE_ISO``
``ISO drive``
``.``
``:MOUNT_TYPE_VIRTUAL``
``internal virtual drive``
``.``

kcgen commented on src/dos/dos.cpp:
```diff
@@ -1426,10 +1427,11 @@ static Bitu DOS_25Handler(void)
 static Bitu DOS_26Handler(void) {
 	LOG(LOG_DOSMISC,LOG_NORMAL)("int 26 called: hope for the best!");
+	std::string type_fat = drive_info.at(DRIVETYPE::FAT);
 	if (reg_al >= DOS_DRIVES || !Drives[reg_al] || Drives[reg_al]->isRemovable()) {	
 		reg_ax = 0x8002;
 		SETFLAGBIT(CF,true);
-	} else if (strncmp(Drives[reg_al]->GetInfo(), "fatDrive", 8) == 0) {
+	} else if (std::string(Drives[reg_al]->GetInfo()).substr(0, type_fat.size()) == type_fat) {
```
I see an enum has been added:

```c++
enum class DRIVETYPE : uint16_t {
	LOCAL   = 1,
	CDROM   = 2,
	FAT     = 3,
	ISO     = 4,
	VIRTUAL = 5,
};
```

And a new global variable, as map, that looks up strings:

```c++
inline std::unordered_map<DRIVETYPE, const char *> drive_info;

drive_info.insert({DRIVETYPE::LOCAL, MSG_Get("MOUNT_TYPE_LOCAL_DIRECTORY")});
drive_info.insert({DRIVETYPE::CDROM, MSG_Get("MOUNT_TYPE_CDROM")});
drive_info.insert({DRIVETYPE::FAT, MSG_Get("MOUNT_TYPE_FAT")});
drive_info.insert({DRIVETYPE::ISO, MSG_Get("MOUNT_TYPE_ISO")});
drive_info.insert({DRIVETYPE::VIRTUAL, MSG_Get("MOUNT_TYPE_VIRTUAL")});
```

kcgen replied:
And these looked up strings are being used through the drive_info logic, just like before.

The goal of using an `enum` is avoid using string to hold type information, and replace the string comparisons with enum comparisons.

kcgen replied:
The `else if (std::string(Drives[reg_al]->GetInfo()).substr(0, type_fat.size()) == type_fat)`

statement shouldn't be doing a string-compare. It should compare enum values.

If `->GetInfo()` previously returned a string-type, we want it to return an DriveType value. For example, `DriveType::Local`.



else if ().substr(0, type_fat.size()) == type_fat)

kcgen replied:
So for the most part, this PR should be "de-string'ifying" the code.

The only part where we want to use the strings (the message strings), is during the logging for the user.

Wengier replied:
@kcgen I did not use enum class that often previously, and many thanks for showing detailed examples of them. I have revised the code accordingly which now use enum comparison instead of string comparison as suggested.

kcgen commented:
Also suggest UpperCamelCase for the enum and it's inside types. DriveaType::Fat

Can use `git grep "enum class" | grep -v libs` to list all the enum classes in the codebase. The majority use this suggested casing (only a few that don't follow that pattern).

Wengier commented:
@kcgen Thanks for the review! I changed `DRIVETYPE` to `DriveType` just before seeing your last message, but now I have changed `LOCAL`=>`Local`, `CDRom`->`Cdrom`, `FAT`=>`Fat`, etc as well (as you tried to show). Even though casing like `Cdrom` may appear to look a bit of strange, I guess it is the desired result.

kcgen commented on include/dos_system.h:
```diff
@@ -305,10 +314,34 @@ class DOS_Drive {
+	{
+		return info;
+	}
+	std::string GetInfoString() const
```
Can return a `const char *`, which is what MSG_Get returns.

Wengier replied:
It is not exactly MSG_Get, but a combination of const char * and std::string for some cases (except virtual and unknown). If changing to const char * may have to use .c_str() for these cases.

kcgen replied:
Ahh ok, I see the `+ info;` additions. :+1:

kcgen commented on include/dos_system.h:
```diff
@@ -305,10 +314,34 @@ class DOS_Drive {
 
 	char curdir[DOS_PATHLENGTH];
 	char info[256];
+	DosDriveType type;
```
Can use default initializers for these members:
```c++
	char curdir[DOS_PATHLENGTH] = {};
	char info[256] = {};
	DosDriveType type = DosDriveType::Unknown;
```

Which should fix the effective C++ warning.

Wengier replied:
Done accordingly. Thanks!

kcgen commented on src/dos/program_mount_common.cpp:
```diff
@@ -85,3 +85,12 @@ void AddCommonMountMessages() {
 	MSG_Add("PROGRAM_MOUNT_STATUS_1", "The currently mounted drives are:\n");
 }
+
+void AddMountTypeMessages() {
+	MSG_Add("MOUNT_TYPE_LOCAL_DIRECTORY", "local directory");
+	MSG_Add("MOUNT_TYPE_CDROM", "CD-ROM drive");
+	MSG_Add("MOUNT_TYPE_FAT", "FAT drive");
+	MSG_Add("MOUNT_TYPE_ISO", "ISO drive");
+	MSG_Add("MOUNT_TYPE_VIRTUAL", "internal virtual drive");
+	MSG_Add("MOUNT_TYPE_UNKOWN", "unknown drive");
+}
```
:+1:
